packages/11ty/_plugins/components/addComponentTag.js

```diff
@@ -19,9 +19,13 @@ module.exports = function(eleventyConfig, component, tagName) {
    * @see https://www.11ty.dev/docs/languages/javascript/#javascript-template-functions
    * @see https://www.11ty.dev/docs/languages/javascript/#relationship-to-filters-and-shortcodes
    */
-  eleventyConfig.addJavaScriptFunction(tagName, function(...args) {
-    return component(eleventyConfig, globalData, { page: this.page })(...args)
-  })
+  component.constructor.name === "AsyncFunction"
+    ? eleventyConfig.addJavaScriptFunction(tagName, async function (...args) {
+        return await component(eleventyConfig, globalData, { page: this.page })(...args)
+      })
+    : eleventyConfig.addJavaScriptFunction(tagName, function (...args) {
+        return component(eleventyConfig, globalData, { page: this.page })(...args)
+      })
```

Comment:
Using an _expression_ without a left-hand assignment might not be parsed well, if moving the ternary to the `return` statement does not work it may read better and be parser safer as you had written it using an `if/else` statement.